drivers: led: is31fl3194.c Add support for blink in RGB and Single modes

anthony289478 opened the pull request:
Add is31fl3194 support for blinking LED in both RGB and Single channel modes

anthony289478 commented:
Would anyone be able to tell me the average lead time for reviewing feature/enhancements?

pillo79 reviewed (commented):
Thanks for improving the driver. A few tiny fixes but otherwise LGTM!

anthony289478 commented on drivers/led/is31fl3194.c:
```diff
@@ -78,22 +98,42 @@ static int is31fl3194_get_info(const struct device *dev,
+		ret = i2c_reg_write_byte_dt(&config->bus, IS31FL3194_CONF_REG, data->conf_reg);
+		if (ret != 0) {
+			return ret;
+		}
```
Updated both L130 and L 305 usages

simonguinot reviewed (changes requested):
Hi @anthony289478,

First, let me apologize for the delay with this review.

I only partially reviewed it because it is a big chunk to chew. And I think it would be helpful to split this patch into several.

Please see my comments below.

anthony289478 committed:
drivers: led: is31fl3194.c Reset state on init

Reset is31fl3194 on startup.
With the exception of power-on-reset, a system reset
does not guarantee the device is in its initial state
so always reset the device on reboot.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

anthony289478 committed:
drivers: led: is31fl3194.c Add support for SDB pin

Add initial support to enable SDB if present to enable output on
all channels on is31fl3194.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

anthony289478 committed:
drivers: led: is31fl3194.c Add enum for RGB and SINGLE modes

Add an enum to identify the current selected DT mode.

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

anthony289478 committed:
drivers: led: is31fl3194.c Add support for blink

Add is31fl3194 support for blinking LED in both
RGB and single channel modes

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

simonguinot reviewed (changes requested):
Hi @anthony289478,

Thanks for splitting this PR into more patches (one per feature / purpose). It's much clearer now.

See below for my comments and questions.

anthony289478 commented on drivers/led/is31fl3194.c:
```diff
@@ -145,11 +161,144 @@ static int is31fl3194_set_color(const struct device *dev, uint32_t led, uint8_t
+
+	/* for both modes, we blink all at the same time to avoid color phase shifts
+	 * which will eventually occur from each channel's internal timers
+	 */
```
If you use the LED's in single channel pattern mode (one pattern timer for each channel) instead of RGB pattern mode (p1 timer for all 3 channels), eventually over several hours you will see that the led's will blink out of order instead of all at once.

simonguinot replied:
Interesting.

Suggestion:

`RGB mode is selected for blinking. Single mode cannot be used, as the LED channels blink out of sync after a few hours.`

anthony289478 commented on drivers/led/is31fl3194.c:
```diff
@@ -295,6 +297,13 @@ static int is31fl3194_init(const struct device *dev)
 		return ret;
 	}
 
+	if (config->gpio_enable.port) {
```
Updated

anthony289478 commented on drivers/led/is31fl3194.c:
```diff
@@ -89,11 +100,11 @@ static int is31fl3194_set_color(const struct device *dev, uint32_t led, uint8_t
 	}
 
-	if (num_colors != 3) {
+	if (num_colors != info->num_colors) {
```
Updated

anthony289478 commented on drivers/led/is31fl3194.c:
```diff
@@ -145,11 +161,144 @@ static int is31fl3194_set_color(const struct device *dev, uint32_t led, uint8_t
+	/* timings are shared between all channels in pattern mode so after setting the color,
+	 * only the first pattern timing registers needs to be set
+	 */
+	if ((data->mode == IS31FL3194_MODE_RGB) && led > 0) {
```
Updated.

anthony289478 commented on drivers/led/is31fl3194.c:
```diff
@@ -145,11 +161,144 @@ static int is31fl3194_set_color(const struct device *dev, uint32_t led, uint8_t
+	}
+
+	if (data->mode == IS31FL3194_MODE_RGB) {
+		for (int i = 0; i < 3; i++) {
```
Updated to only call on led 0

anthony289478 commented:
> Pipeline is failing twice with this error.
> 
> ```
> 2025-09-24T15:10:51.6030716Z ModuleNotFoundError: No module named 'jsonschema'
> ```
> 
> If someone with permissions could return the pipeline when actions are fixed I would appreciate it.

Rebased branch from [8d05603](https://github.com/zephyrproject-rtos/zephyr/commit/8d05603927bb3c864ce47e11af050e31bab55447) to [3958a69](https://github.com/zephyrproject-rtos/zephyr/commit/3958a69dae71ca1ce195d9323d56a0b2e44c10d6) latest main to fix pipeline

simonguinot commented on drivers/led/is31fl3194.c:
```diff
@@ -145,11 +161,144 @@ static int is31fl3194_set_color(const struct device *dev, uint32_t led, uint8_t
+	uint8_t value;
+	uint8_t address;
+	uint8_t base;
+
```
Well. I think "forever" means it doesn't stop after a number of blinks. But yes, user need a way to configure its LED and then to disable a blinking.

Unfortunately the method is not defined and the behavior is not the same among the LED drivers. If we take the `led_pwm` driver for example, changing the brightness configuration stops the blinking. The `ht16k33` driver stops the blinking is `delay_off` is set to 0. Other drivers doesn't seem to stop at all... So there is no rule.

Here is what I suggest:

- `delay_off`=0, `delay_on`=x: LED on and blinking disabled
- `delay_off`=x, `delay_on`=0: LED off and blinking disabled
- `delay_off`=0, `delay_on`=0: it is a special case... An idea could be to use it as an "auto" blinking mode with the period set to a default value, pleasant for user and convenient for the hardware. Or it could be LED off...

simonguinot commented:
> FYI I will be on vacation starting tomorrow for the next two weeks and will not be able to update in response to change requests.

Noted ! Enjoy your vacation :)

anthony289478 committed:
drivers: led: is31fl3194.c Add support to disable blink

Add is31fl3194 support to set led output state while
disabling blinking with specific arguments into led_blink().

delay_off=0, delay_on=x: LED on and blinking disabled
delay_off=x, delay_on=0: LED off and blinking disabled

Signed-off-by: Anthony Williams <anthony289478@gmail.com>

simonguinot commented:
Hi @anthony289478. Seen. I'll try to review it this week.
